src/libraries/System.Text.RegularExpressions/gen/RegexAnalyzer.cs

```diff
@@ -0,0 +1,416 @@
+        {
+            // Ensure source generator is supported.
+            if (!ProjectSupportsRegexSourceGenerator(context.Operation))
+            {
+                return;
+            }
+
+            // Ensure the object creation is a call to the Regex constructor.
+            IObjectCreationOperation operation = (IObjectCreationOperation)context.Operation;
+            if (!SymbolEqualityComparer.Default.Equals(operation.Type, context.Compilation.GetTypeByMetadataName(RegexTypeName)))
+            {
+                return;
+            }
+
+            // Ensure that all inputs to the constructor are constant.
+            foreach (IArgumentOperation argument in operation.Arguments)
+            {
+                if (!IsConstant(argument))
+                {
+                    return;
+                }
+            }
+
+            // Create the property bag.
+            ImmutableDictionary<string, string?> properties = ImmutableDictionary.CreateRange(new[]
+            {
+                new KeyValuePair<string, string?>(PatternIndexName, "0"),
+                new KeyValuePair<string, string?>(RegexOptionsIndexName, operation.Arguments.Length > 1 ? "1" : null),
+                new KeyValuePair<string, string?>(RegexTimeoutIndexName, operation.Arguments.Length > 2 ? "2" : null),
+                new KeyValuePair<string, string?>(RegexTimeoutName, operation.Arguments.Length > 2 ? CalculateMillisecondsFromTimeSpan(operation.Arguments[2].Value)?.ToString(CultureInfo.InvariantCulture) : null),
+            });
+
+            // Report the diagnostic.
+            SyntaxNode? syntaxNodeForDiagnostic = operation.Syntax.ChildNodes().FirstOrDefault()!.Parent;
+            Debug.Assert(syntaxNodeForDiagnostic is not null);
+            context.ReportDiagnostic(Diagnostic.Create(DiagnosticDescriptors.UseRegexSourceGeneration, syntaxNodeForDiagnostic.GetLocation(), properties));
+        }
+
+        /// <summary>
+        /// Calculates the constant total milliseconds represented by a timespan operation.
+        /// </summary>
+        /// <remarks>
+        /// RegEx constructors and static methods optionally take a <see cref="TimeSpan"/> parameter to represent when it should timeout to avoid
+        /// catastrophic backtracking, but the RegexGeneratorAttribute can't take a non-constant parameter so it instead takes an optional int milliseconds.
+        /// This method is in charge of verifying if that timespan parameter has a constant value, and if it does, it makes the transformation from Timespan
+        /// to milliseconds.
+        /// </remarks>
+        /// <param name="operation">The operation representing the parameter.</param>
+        /// <returns>The total milliseconds in case the TimeSpan value is constant, or <see langword="null"/> if there is no constant value.</returns>
+        private static int? CalculateMillisecondsFromTimeSpan(IOperation operation)
```

Comment:
We're going to need _a lot_ of tests around this. :)  We could also consider just giving up on the diagnostic in the case where a timeout is specified.

Reply:
Decided to remove this support for now.